File: 김은지/주식가격.js
// function solution(prices) {
//   const afterPrices = [];
//   const terms = [];

//   for (let i = prices.length - 1; i >= 0; i--) {
//     let term = 0;
//     for (let j = afterPrices.length - 1; j >= 0; j--) {
//       term++;
//       if (prices[i] > afterPrices[j]) break;
//     }
//     terms.push(term);
//     afterPrices.push(prices[i]);
//   }

//   return terms.reverse();
// }

function solution(prices) {
  const durations = Array(prices.length).fill(0);
  const waitingPriceIdx = [];

  prices.forEach((currPrice, currentIdx) => {
    while (waitingPriceIdx.length > 0 && prices[waitingPriceIdx.at(-1)] > currPrice) {
      const prevIdx = waitingPriceIdx.pop();
      durations[prevIdx] = currentIdx - prevIdx;
    }
    waitingPriceIdx.push(currentIdx);
  });

  while (waitingPriceIdx.length > 0) {
    const idx = waitingPriceIdx.pop();
    durations[idx] = prices.length - 1 - idx;
  }

  return durations;
}

File: 김현경/주식가격.py
def solution(prices):
    answer = [0] * len(prices)
    stack = []
    
    for i in range(len(prices)):
        while stack and prices[i] < prices[stack[-1]]:
            idx = stack.pop()
            answer[idx] = i - idx
        stack.append(i)
    
    while stack:
        idx = stack.pop()
        answer[idx] = len(prices) - idx - 1
        
    return answer

# 처음 풀었던 코드
# def solution(prices):
#     answer = [0] * len(prices)
    
#     for i in range(len(prices)):
#         for j in range(i+1, len(prices)):
#             if prices[i] > prices[j]:
#                 answer[i] += 1
#                 break
#             else:
#                 answer[i] += 1
#     return answer